### ISSUE TITLE
hlint v2.2.11 not uploaded to hackage

### ISSUE THREAD

**Issue by essandess** [ndmitchell/hlint#951]
I'm trying to update MacPorts `hlint` to v2.2.11. We use hackage as the source, but 2.2.11 hasn't been uploaded yet. Would you please upload released hlint versions to hackage?

Related:
* https://github.com/haskell/cabal/issues/6699
* https://github.com/macports/macports-ports/pull/6182

**Comment by ndmitchell** [comment]
As far as I can tell it's there: http://hackage.haskell.org/package/hlint-2.2.11 - the way I release is uploading to Hackage, so if it's not there, it's not released :)

**Comment by essandess** [comment]
Whoops of course it is.

I'm failing to build `hlint` 2.2.11 on macOS using `cabal` 3.2.0.0 and `ghc` 8.10.1, and getting these errors.

Would you please tell me which tree you believe I should be barking up? `hlint`? `cabal`? `ghc`?

```
:debug:configure system:  cd "/opt/local/var/macports/build/_opt_local_ports_devel_hlint/hlint/work/hlint-2.2.11" && /opt/local/bin/cabal new-configure --prefix=/opt/local --bindir=/opt/local/bin --libdir=/opt/local/lib --libsubdir=hlint --dynlibdir=/opt/local/lib --libexecdir=/opt/local/libexec --libexecsubdir=hlint --datadir=/opt/local/share/hlint --docdir=/opt/local/share/doc/hlint --htmldir=/opt/local/share/doc/hlint --sysconfdir=/opt/local/etc/hlint --enable-documentation --enable-relocatable
:info:configure Resolving dependencies...
:info:configure cabal: Could not resolve dependencies:
:info:configure [__0] trying: hlint-2.2.11 (user goal)
:info:configure [__1] next goal: ghc-lib-parser (dependency of hlint +/-ghc-lib)
:info:configure [__1] rejecting: ghc-lib-parser-8.10.1.20200324 (conflict: hlint +/-ghc-lib =>
:info:configure ghc-lib-parser==8.8.*)
:info:configure [__1] trying: ghc-lib-parser-8.8.3.20200224
:info:configure [__2] next goal: base (dependency of hlint)
:info:configure [__2] rejecting: base-4.14.0.0/installed-4.14.0.0 (conflict: ghc-lib-parser =>
:info:configure base>=4.11 && <4.14)
:info:configure [__2] rejecting: base-4.12.0.0, base-4.11.1.0, base-4.11.0.0, base-4.10.1.0,
:info:configure base-4.10.0.0, base-4.9.1.0, base-4.9.0.0, base-4.8.2.0, base-4.8.1.0,
:info:configure base-4.8.0.0, base-4.7.0.2, base-4.7.0.1, base-4.7.0.0, base-4.6.0.1,
:info:configure base-4.6.0.0, base-4.5.1.0, base-4.5.0.0, base-4.4.1.0, base-4.4.0.0,
:info:configure base-4.3.1.0, base-4.3.0.0, base-4.2.0.2, base-4.2.0.1, base-4.2.0.0,
:info:configure base-4.1.0.0, base-4.0.0.0, base-3.0.3.2, base-3.0.3.1 (constraint from
:info:configure non-upgradeable package requires installed instance)
:info:configure [__2] fail (backjumping, conflict set: base, ghc-lib-parser, hlint)
:info:configure non-upgradeable package requires installed instance)
:info:configure [__2] fail (backjumping, conflict set: base, ghc-lib-parser, hlint)
:info:configure After searching the rest of the dependency tree exhaustively, these were the
:info:configure goals I've had most trouble fulfilling: hlint, base, ghc-lib-parser
```

**Comment by essandess** [comment]
Please see this comment from @chshersh. Apparently there is a discrepancy between the hackage and GitHub releases:

https://github.com/haskell/cabal/issues/6699#issuecomment-612606025
> @essandess I see there was some confusion. `hlint-2.2.11` is indeed uploaded to Hackage but the `2.2.11` version is not compatible with GHC-8.10 according to its dependencies. I can see from the Hackage dependencies for `hlint` the following constraints:
> 
> ```
> ghc == 8.8.*
> ghc-lib-parser == 8.8.*
> ```
> 
> I advised using the version from GitHub that is compatible with GHC 8.10 according to its code in the `hlint.cabal` file:
> 
> ```cabal
>     if !flag(ghc-lib) && impl(ghc >= 8.10.0) && impl(ghc < 8.11.0)
>       build-depends:
>         ghc == 8.10.*,
>         ghc-boot-th,
>         ghc-boot
>     else
>       build-depends:
>           ghc-lib-parser == 8.10.*
>     build-depends:
>         ghc-lib-parser-ex >= 8.10.0.3 && < 8.10.1
> ```
> 
> Although the value of the `version` field on GitHub is the same — 2.2.11 — the content of the package on GitHub is not the same as on Hackage. So the actual version of `hlint` on GitHub is not 2.2.11, it's something different. I can see how this can be confusing. But this is a typical development workflow for Haskell packages:
> 
> 1. You release a version to Hackage.
> 2. You continue working on your project, implementing features without changing the version in your `.cabal` file.
> 3. Before releasing to Hackage, you evaluate all your changes locally and assign a new version according to your changes, following either PVP or SemVer versioning schemes (or using [policeman](https://github.com/kowainik/policeman) to detect the new version automatically)

**Comment by googleson78** [comment]
I don't see a discrepancy - the latest github release is also on `8.8`.

Also wouldn't having an `8.10` still **not** fix your issues, seeing as how `8.10` [is not in macports yet](https://github.com/macports/macports-ports/pull/6849)?

**Comment by shayne-fletcher** [comment]
Looking at the failure I see that `base 4.14` is out-of-bounds for `ghc-lib-parser`, that's the problem.

**Comment by shayne-fletcher** [comment]
Have opened https://github.com/digital-asset/ghc-lib/issues/194 to track this. If I can get this fixed with a new ghc-lib-8.8.3 release to Hackage, hlint will automatically pick it up. Hence suggest this ticket be closed @ndmitchell .

**Comment by shayne-fletcher** [comment]
@essandess `ghc-lib-8.8.3.20200412` is now on hackage. You should be good to go.

**Comment by essandess** [comment]
> I don't see a discrepancy - the latest github release is also on `8.8`.
> 
> Also wouldn't having an `8.10` still **not** fix your issues, seeing as how `8.10` [is not in macports yet](https://github.com/macports/macports-ports/pull/6849)?

Thanks all for the help.

I have a MacPorts PR open for ghc v8.10.1 and am trying to build local dependencies to bring everything up-to-date; MacPorts will be on 8.10.1 very soon.

**Comment by essandess** [comment]
@shayne-fletcher Now more dependency issues with `ghc-lib-parser`, looks like `ghc-prim`.

Also, in the future, where’s the appropriate place to open a dependency issues like this?

Log:
```
:debug:configure system:  cd "/opt/local/var/macports/build/_opt_local_ports_devel_hlint/hlint/work/hlint-2.2.11" && /opt/local/bin/cabal new-configure --prefix=/opt/local --bindir=/opt/local/bin --libdir=/opt/local/lib --libsubdir=hlint --dynlibdir=/opt/local/lib --libexecdir=/opt/local/libexec --libexecsubdir=hlint --datadir=/opt/local/share/hlint --docdir=/opt/local/share/doc/hlint --htmldir=/opt/local/share/doc/hlint --sysconfdir=/opt/local/etc/hlint --enable-documentation --enable-relocatable
:info:configure Resolving dependencies...
:info:configure cabal: Could not resolve dependencies:
:info:configure [__0] trying: hlint-2.2.11 (user goal)
:info:configure [__1] next goal: ghc-lib-parser (dependency of hlint +/-ghc-lib)
:info:configure [__1] rejecting: ghc-lib-parser-8.10.1.20200324 (conflict: hlint +/-ghc-lib =>
:info:configure ghc-lib-parser==8.8.*)
:info:configure [__1] trying: ghc-lib-parser-8.8.3.20200412
:info:configure [__2] next goal: base (dependency of hlint)
:info:configure [__2] rejecting: base-4.14.0.0/installed-4.14.0.0 (conflict: ghc-lib-parser =>
:info:configure ghc-prim>0.2 && <0.6, base => ghc-prim==0.6.1/installed-0.6.1)
:info:configure [__2] rejecting: base-4.12.0.0, base-4.11.1.0, base-4.11.0.0, base-4.10.1.0,

:info:configure base-4.10.0.0, base-4.9.1.0, base-4.9.0.0, base-4.8.2.0, base-4.8.1.0,
:info:configure base-4.8.0.0, base-4.7.0.2, base-4.7.0.1, base-4.7.0.0, base-4.6.0.1,
:info:configure base-4.6.0.0, base-4.5.1.0, base-4.5.0.0, base-4.4.1.0, base-4.4.0.0,
:info:configure base-4.3.1.0, base-4.3.0.0, base-4.2.0.2, base-4.2.0.1, base-4.2.0.0,
:info:configure base-4.1.0.0, base-4.0.0.0, base-3.0.3.2, base-3.0.3.1 (constraint from
:info:configure non-upgradeable package requires installed instance)
:info:configure [__2] fail (backjumping, conflict set: base, ghc-lib-parser, hlint)
:info:configure After searching the rest of the dependency tree exhaustively, these were the
:info:configure goals I've had most trouble fulfilling: hlint, base, ghc-lib-parser
```

**Comment by shayne-fletcher** [comment]
> Also, in the future, where’s the appropriate place to open a dependency issues like this?

https://github.com/digital-asset/ghc-lib/issues

**Comment by shayne-fletcher** [comment]
@essandess Let's move the conversation to https://github.com/digital-asset/ghc-lib/issues/194 please.

**Comment by shayne-fletcher** [comment]
@essandess I'm flying blind since I don't have the capacity to build with 8.10 yet (waiting on a stack resolver). If I were to give you a .tar.gz to test, would you be able to do that?

**Comment by essandess** [comment]
Moved to https://github.com/digital-asset/ghc-lib/issues/194

**Comment by ndmitchell** [comment]
If the latest released HLint won't build with the latest released GHC, that's an HLint bug. It might _also_ be a bug in ghc-lib or similar, but I'll always take a bug in HLint too. Just verifying with a recent cabal update and a brand new ghc-8.10.

**Comment by essandess** [comment]
@ndmitchell `hlint` build perfectly now thanks to the pushes by @shayne-fletcher. Thanks again to everyone for the help. I’ll update MacPorts with the latest. The only thing for us left is a bootstrapped `cabal`, but that’s a separate issue.

**Comment by ndmitchell** [comment]
Awesome! Thanks @essandess for reporting and tracking it down, and thanks @shayne-fletcher for fixing it.
